Code/Framework/AzCore/AzCore/Name/Name.cpp

```diff
@@ -44,39 +46,51 @@ namespace AZ
         *this = rhs;
     }
 
+    Name::Name(NameRef name)
+        : m_data(AZStd::move(name))
+        , m_hash(name->GetHash())
+        , m_view(name->GetName())
+    {
+    }
+
+    Name Name::FromStringLiteral(AZStd::string_view name)
+    {
+        Name literalName;
+        literalName.SetNameLiteral(name);
+        return literalName;
+    }
+
     Name& Name::operator=(const Name& rhs)
     {
-        m_data = rhs.m_data;
-        m_hash = rhs.m_hash;
-        if (!rhs.m_view.empty())
+        if (rhs.m_supportsDeferredLoad && rhs.m_data == nullptr)
         {
-            m_view = rhs.m_view;
+            m_hash = rhs.m_hash;
+            m_data = rhs.m_data;
+            SetNameLiteral(rhs.GetStringView());
         }
         else
         {
-            SetEmptyString();
+            m_data = rhs.m_data;
+            m_hash = rhs.m_hash;
+            m_view = rhs.m_view;
         }
         return *this;
     }
 
     Name& Name::operator=(Name&& rhs)
     {
-        if (rhs.m_view.empty())
+        if (rhs.m_supportsDeferredLoad)
```

Comment:
I'm erring on the side of preserving the deferred support, so that this new `Name` can survive the dictionary being destroyed and recreated - this actually happens during `AZ_NAME_LITERAL` because we use `Name::FromStringLiteral` to create the `static Name`, and function-scope static initialization can occur while the dictionary is alive - the dictionary can still be subsequently destroyed and recreated, so we have these new literal Names support that as well.

Reply:
Why do we want to support moving statically initialized names anyway? It seems like something that would be reasonable to assert against and disallow.

Reply:
`FromStringLiteral` creates a new name that we'll want to store with a move operation, so anything like `static AZ::Name foo = AZ::Name::FromStringLiteral("foo");` needs this.

Reply:
I think I'm getting hung up on: what does it actually mean to move a name that supports deferred load? To me it would mean you have to transfer over all the data, including its position in the linked list, completely invalidating the original Name. The new Name should take the old Name's place in the linked list. Anything else leaves me pretty confused because it doesn't follow the typical metaphor of moving data. But maybe that's not possible to do safely because what if the move is across module boundaries?

Part of my confusion might come from the fact that rhs.m_supportsDeferredLoad doesn't necessarily mean the Name is in the static linked list (see my other comment below).

Reply:
Generally you shouldn't be moving the name literal across module boundaries, this exists primarily to let From string literal have its result get moved. If you do decide to move a name literal you're assuming the lifespan of its view, whatever that might be - as a best practice I'd say only ever copy literals you define though